### DISCUSSION TITLE
[OPi5] Is it possible to get the real time CPU frequency?

### DISCUSSION THREAD

**Discussion by nobara256** [Joshua-Riek/ubuntu-rockchip#173] [Q&A] [answered]
Device: Orange Pi 5 
Image: v1.16

By far I have tested every combination of commands that I can find on the internet to query the current CPU frequency. They only seem to show the MAX frequency of the cores but not the real-time frequencies.

Also, do the A76 cores ever hit their max ~2.6Ghz or they are capped at 1.8Ghz like their little brothers?

**Comment by Joshua-Riek** [comment]
You should be able to view the current CPU frequency using the sysfs interface.
```bash
ubuntu@ubuntu-desktop:~$ cat /sys/bus/cpu/devices/cpu*/cpufreq/scaling_cur_freq
1800000
1800000
1800000
1800000
2304000
2304000
2304000
2304000
```
I see the A55 cores are at 1.8Ghz and the A76 cores are around ~2.3Ghz.

**Reply by nobara256** [reply]
Yeah, they all report the MAX frequency of the cores. I wonder if all of them running at their MAX frequencies all the time or they slow down when there is no load.

**Reply by Joshua-Riek** [reply, marked as answer]
Right, this is the expected behavior by default. I have two systemd services that set the performance governor for the GPU and CPU to `performance` instead of the default `ondemand` value. This is because I noticed the CPU and GPU are slightly delayed when increasing the frequency under load, causing stuttering. 

If you want to change only the cpu governor yourself to ondemand, you can use this command.
```bash
echo ondemand | sudo tee /sys/bus/cpu/devices/cpu*/cpufreq/scaling_governor
```

One example I observed a while ago is dragging the Chromium window around. It can stutter and become laggy when the governor is set to its default value of ondemand. So I decided to have two services set the CPU and GPU to max performance out of the box. 

Both services can be disabled easily with the below commands and a reboot.
```bash
sudo systemctl disable cpu-governor-performance
sudo systemctl disable gpu-governor-performance
```

**Reply by nobara256** [reply]
Thanks for the explanation.

**Reply by bryankanderson** [reply]
Thank you!  Building a portable media server for trips, etc.  So amazed with this platform!

**Comment by guynich** [comment]
on Ubuntu I'm using python and the `psutil` [package](https://github.com/giampaolo/psutil) `cpu_freq()` method to see real time CPU frequencies on RK3588.

**Reply by MarkhamLee** [reply]
I use the psutil package as well to get real time per CPU frequencies on my Orange Pi 5 plus. However, whenever I run: freq = psutil.cpu_freq(percpu=True) I just get three frequencies instead of one for each of the eight cores. E.g., - [scpufreq(current=1800.0, min=1800.0, max=1800.0), scpufreq(current=2352.0, min=1200.0, max=2352.0), scpufreq(current=2352.0, min=408.0, max=2352.0)]. I'm guessing it's similar to how the psutil.sensors_temperatures() method gives you temps for bigcore0, bigcore1 and littlecore - i.e. the library is only seeing three cores. I suspect this is related to the library itself and how it deals with big/little core configurations, as I've noticed similarly odd behavior/numbering on my Intel machines with the big/little setup.

**Reply by guynich** [reply]
I think the three CPU clusterings are enumerated here:
`ls /sys/devices/system/cpu/cpufreq/policy*`
e.g.:
`cat /sys/devices/system/cpu/cpufreq/policy0/affected_cpus`  # littlecore 0-3.
`cat /sys/devices/system/cpu/cpufreq/policy4/affected_cpus`  # bigcore 4-5.
`cat /sys/devices/system/cpu/cpufreq/policy5/affected_cpus`  # bigcore 6-7.
